From 484e1412f7a9711e6e928aa9692ae9d9d70fc9f0 Mon Sep 17 00:00:00 2001
From: "Kurtadkar, Yashodhan" <yashodhan.kurtadkar@intel.com>
Date: Fri, 27 Feb 2015 14:35:41 +0530
Subject: [PATCH] BQ24192: Update VINDPM settings

VINDPM values need to be updated to maintain sufficient gap between
Vin and VBAT. Without proper DPM settings, the charger will reduce
the charge current, and there could be input collapse if the VBAT
is close to DPM settings. Since the charger always looks into VBAT
(not the OCV), driver should look into VOLTAGE_NOW property rather
than VOLTAGE_OCV.

Change-Id: I01f0afccd110295551e479ef723f5650fa27fa62
Tracked-On: https://jira01.devtools.intel.com/browse/AKP-1118
Signed-off-by: Kurtadkar, Yashodhan <yashodhan.kurtadkar@intel.com>
---
 drivers/power/bq24192_charger.c | 17 +++++++++++------
 1 file changed, 11 insertions(+), 6 deletions(-)

diff --git a/drivers/power/bq24192_charger.c b/drivers/power/bq24192_charger.c
index ad958fa..7383e92 100644
--- a/drivers/power/bq24192_charger.c
+++ b/drivers/power/bq24192_charger.c
@@ -65,12 +65,14 @@
 #define INPUT_SRC_VOLT_LMT_404                 (1 << 4)
 #define INPUT_SRC_VOLT_LMT_412                 (3 << 3)
 #define INPUT_SRC_VOLT_LMT_444                 (7 << 3)
+#define INPUT_SRC_VOLT_LMT_460                 (9 << 3)
 #define INPUT_SRC_VOLT_LMT_468                 (5 << 4)
 #define INPUT_SRC_VOLT_LMT_476                 (0xB << 3)
 
 #define INPUT_SRC_VINDPM_MASK                  (0xF << 3)
 #define INPUT_SRC_LOW_VBAT_LIMIT               3600
-#define INPUT_SRC_MIDL_VBAT_LIMIT              4100
+#define INPUT_SRC_MIDL_VBAT_LIMIT1             3800
+#define INPUT_SRC_MIDL_VBAT_LIMIT2             4100
 #define INPUT_SRC_HIGH_VBAT_LIMIT              4350
 
 /* D0, D1, D2 represent the input current limit */
@@ -1835,7 +1837,7 @@ static void bq24192_task_worker(struct work_struct *work)
 	/* Modify the VINDPM */
 
 	/* read the battery voltage */
-	vbatt = fg_chip_get_property(POWER_SUPPLY_PROP_VOLTAGE_OCV);
+	vbatt = fg_chip_get_property(POWER_SUPPLY_PROP_VOLTAGE_NOW);
 	if (vbatt == -ENODEV || vbatt == -EINVAL) {
 		dev_err(&chip->client->dev, "Can't read voltage from FG\n");
 		goto sched_task_work;
@@ -1851,11 +1853,14 @@ static void bq24192_task_worker(struct work_struct *work)
 	 * Hence disabling dynamic vindpm update  for bq24296 chip.
 	*/
 	if (chip->chip_type != BQ24296) {
-		if (vbatt >= INPUT_SRC_MIDL_VBAT_LIMIT &&
-			vbatt <= INPUT_SRC_HIGH_VBAT_LIMIT)
-			vindpm = INPUT_SRC_VOLT_LMT_444;
-		else
+		if (vbatt > INPUT_SRC_LOW_VBAT_LIMIT &&
+			vbatt <= INPUT_SRC_MIDL_VBAT_LIMIT1)
 			vindpm = INPUT_SRC_VOLT_LMT_412;
+		else if (vbatt > INPUT_SRC_MIDL_VBAT_LIMIT1 &&
+			vbatt <= INPUT_SRC_MIDL_VBAT_LIMIT2)
+			vindpm = INPUT_SRC_VOLT_LMT_444;
+		else if (vbatt > INPUT_SRC_MIDL_VBAT_LIMIT2)
+			vindpm = INPUT_SRC_VOLT_LMT_460;
 
 		mutex_lock(&chip->event_lock);
 		ret = bq24192_modify_vindpm(vindpm);
-- 
1.9.1

